### DISCUSSION TITLE
Alpine pollutes the target element with the properties of `xbind` when the component specified in `x-data` is not defined

### DISCUSSION THREAD

**Discussion by northern-mindset** [alpinejs/alpine#3603] [7. Bugs (archived)] [answered]
```html
<html>
  <head>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
  </head>
  <body>
    <div x-data="component" x-bind="self"></div>
  </body>
</html>
```

Alpine pollutes the `div` element with (I guess) the `window`’s properties (due to the fact that [`self` is a property of `window`](https://developer.mozilla.org/en-US/docs/Web/API/Window/self) and is accessible as a global object):

![image](https://github.com/alpinejs/alpine/assets/6065071/49e10842-194d-4501-8454-3baefee41151)

And consequently a lot of errors in the console:

![image](https://github.com/alpinejs/alpine/assets/6065071/7c602b08-5674-4776-8816-2a297835841f)

This happens when `x-bind` contains a globally accessible variable, such as `self`, `window`, `document`. Changing `self` to `myself`, for example, helps, and then we get only the expected “component is not defined” error along with `myself is not defined` and `can't convert undefined to object`:

![image](https://github.com/alpinejs/alpine/assets/6065071/01772dff-1d2d-439a-a4ae-6c5de3b38af7)

It seems that Alpine mistakenly tries to evaluate `x-bind` even when the component in `x-data` is not defined. I believe this is a bug. 

Regarding the fact that it adds all the properties from an object specified in `x-bind` to the target element, after reading the [documentation](https://alpinejs.dev/directives/bind#bind-directives) again and carefully, I see that it is the expected behavior:

> The object keys can be anything you would normally write as an attribute name in Alpine. This includes Alpine directives and modifiers, **but also plain HTML attributes.**

However, the example in the docs undoubtedly indicates that the object is taken from the component and not from the global scope. Though, it is not stated in the docs explicitly.

**Comment by ekwoka** [comment, marked as answer]
This is expected behavior.

To allow expressions to work without needing to do `this.stuff` or `$data.stuff` the `with` keyword is used which exposes all the values on the data context as if they are defined in scope.

And then the expressions are evaluates in that scope.

Inherently, just like JS, if you were to do `console.log(self)` without a scoped variable called `self`, then `globalThis.self` is what you would get.

This is a fundamental aspect of JavaScript and there isn't any way around that.

Anything that stopped this from happening would also stop `console.log` from working in ane expression.

It's never called out in the docs because it's EXACTLY what everyone would expect to happen. That they have access to the `globalThis`.

**Reply by northern-mindset** [reply]
Thank you for a detailed explanation. I fetched some new knowledge from the docs regarding the scoping and `with`.

However, the `with` keyword seems to [be deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with). Does the team plan to migrate to some other approach eventually?

Whereas it is quite expected to have access to all the global objects in callbacks, such as `x-on` and `x-init`, but I dare to disagree that **everyone** would expect reaching `globalThis` instead of my local data in `x-bind`, `x-show` etc… At least there is me who was not expecting this :) Of course, there is a probability that I'm unique in this, but I doubt it. After your explanation I understand and kind of accept this behavior, but it is damn not obvious from the documentation at all that I can go that wild and spread the whole `window` into my element… Personally, I'd even prefer to write `this.stuff` to avoid this accidental and unwanted evaluation.

I read the documentation quite thoroughly and don't remember any code example where anything outside the component's data is referenced in expressions, as well as any mentioning of `globalThis` etc. Therefore, I friendly suggest mentioning this explicitly. Somewhere in https://alpinejs.dev/essentials/state, https://alpinejs.dev/directives/data#scope, or in the Advanced section.

Besides this, there is another aspect of the issue:

> It seems that Alpine mistakenly tries to evaluate `x-bind` even when the component in `x-data` is not defined. I believe this is a bug.

What do you think about it? Would everyone also expect evaluating all the directives despite the fact that the main directive `x-data` is invalid? [Everything in Alpine starts with the `x-data` directive.](https://alpinejs.dev/directives/data)

Please excuse me if I'm too insistent, I just want to contribute to the project sincerely.

**Reply by ekwoka** [reply]
There isn't really any option to get what Alpine does without `with`.

Or at least none that are anywhere near as performant and consistent.

**Reply by SimoTod** [reply]
Is there a reason why you would expect it not to have access to the global scope? This is just the way javascript works: https://codepen.io/SimoTod/pen/oNaKvVL
The majority of people would hate having to write `x-text="this.foo"` instead of `x-text="foo"`.
The project initially aimed at replicating vue because of its good dev experience and the way you refer to variables is one of the things.

**Reply by northern-mindset** [reply]
> There isn't really any option to get what Alpine does without `with`.

All right. This means that we all hope that browsers will not drop this deprecated feature…

> Is there a reason why you would expect it not to have access to the global scope?

The main reason for this is that it is not documented. As I said above, **none** of the examples in the docs demonstrate referencing global scope, only properties from `x-data`. I'm not trying to fight against JS or Alpine's principles. I'm just requesting for a clear documentation on this point. Don't expect that an average user will automatically know how Alpine is evaluating expressions.

Moreover, I'm not sure if `x-show="window.someProperty"` will be reactive. Again, due to the lack of documentation.

Lastly, let me attract your attention for the third time to the most important aspect:

> It seems that Alpine mistakenly tries to evaluate `x-bind` even when the component in `x-data` is not defined. I believe this is a bug.

**Reply by ekwoka** [reply]
Other directives inside an x-data will still evaluate if the x-data errors.

Alpine can't know if the expressions depend on that x-data or not until it evaluates them.

`With` does not appear to be going away, considering how dedicated to reverse compatibility ECMA and the browsers are.

It is even present in newer runtimes that have much less obligation to support it.

If such a time happens, absolution might be possible, but it would likely necessitate a performance and/or bundle size hit.

Sure, it's not the best, but there isn't really any alternative.

And yes, the docs don't really indicate it, but they never suggest using this. It's why the magic `$data` exists.

The functions are run with the `globalThis`. They are not bound to anything.


And no properties are reactive if they are not inside a reactive object (created by `Alpine.reactive` which the result of data functions does).

That is a bit more of a catch, but should make sense to most anyone pretty quickly if they hit it.

**Reply by northern-mindset** [reply]
Makes sense, thank you for your thoughts.

All right, let's consider that I faced an edge case, ran down a rabbit hole and found an answer here with your help :) Let this discussion be a documentation for future truth-seekers.
